### DISCUSSION TITLE
Problems with loading images from FLASH memory

### DISCUSSION THREAD

**Discussion by grados73** [RudolphRiedel/FT800-FT813#83] [Q&A]
Hello,
I'm using STM32F407 and RVT50HQBNWC00-B from riverdi.
I tried to use your library in my project and examples work fine. Moreover, I can load own image to RAM_G and print them from RAM_G, but I can not print image from external flash which is in display. I don't know if the problem is with my using EVEAB or with my using your library. 
I read https://github.com/RudolphRiedel/FT800-FT813/discussions/36 and make printing from flash almost the same, but it doesn't work.
On the screen doesn't show anything from flash. I was trying to read an image from flash to RAM_G and print them on screen from RAM_G but only about 10% of picture are loading. I was trying to do it in "normal" mode and burst one. 

Not only that, but I have some questions:
-Should I use compressed .bin files or .raw to upload on FLASH?
-Why user pink-fr3ud in discussion #36 use dividing by 32 the flash address:
 _EVE_cmd_setbitmap_burst(0x800000 | (gauge.addressFlash **/ 32**), gauge.format, gauge.w, gauge.h);_

Let me present my action and code, maybe you or someone else will find a mistake in it: 

In EVEAB I'm uploading two images .jpg 304x304 and 200x120, once without compression and once with compression, using format COMPRESSED_RGBA_ASTC_8x8_KHR:
![image](https://user-images.githubusercontent.com/71728314/229493588-f71ed913-38a9-4930-af48-595205e7e90c.png)

In EVAB I'm generating flash image:
![image](https://user-images.githubusercontent.com/71728314/229497456-a5339b1b-6301-4ecd-bd79-5a7456d626a7.png)

.edf file:
```

Name                                                       : ID : Offset : Size  : Compression : Type : Sub-Type : Width : Height

unified.blob                                               : 0  : 0      : 4096  : 0           : 1    : 0        : 0     : 0     
Main_304x304_304x304_COMPRESSED_RGBA_ASTC_8x8_KHR.raw      : 1  : 4096   : 23104 : 0           : 2    : 37815    : 304   : 304   
Main_mini_200x120_200x120_COMPRESSED_RGBA_ASTC_8x8_KHR.raw : 2  : 27200  : 6016  : 0           : 2    : 37815    : 200   : 120   
Main_304x304_304x304_COMPRESSED_RGBA_ASTC_8x8_KHR.bin      : 3  : 33216  : 6400  : 1           : 2    : 37815    : 304   : 304   
Main_mini_200x120_200x120_COMPRESSED_RGBA_ASTC_8x8_KHR.bin : 4  : 39616  : 3584  : 1           : 2    : 37815    : 200   : 120   
```

In EVAB I'm programming flash using button "UPDATE&VERIFY"
![image](https://user-images.githubusercontent.com/71728314/229497558-9a491c37-9ba9-4454-b80d-5487733bd3ce.png)

And now I'm trying to display it on screen using this code:

```
#include "EVE.h"
#include "tft_data.h"
#include "tft.h"

#define TEST_UTF8 0

/* some pre-definded colors */
#define RED     0xff0000UL
#define ORANGE  0xffa500UL
#define GREEN   0x00ff00UL
#define BLUE    0x0000ffUL
#define BLUE_1  0x5dade2L
#define YELLOW  0xffff00UL
#define MAGENTA 0xff00ffUL
#define PURPLE  0x800080UL
#define WHITE   0xffffffUL
#define BLACK   0x000000UL

/* memory-map defines */
#define MEM_FONT 0x000f7e00 /* the .xfont file for the UTF-8 font is copied here */
#define MEM_LAWETA_KW 0x000f8000 /* start-address of logo, needs 6 272 bytes of memory. 0x000f8000 = 1 015 808(DEC)*/

#define MEM_LAWETA_MINI 0x000FD980

#define MEM_LAWETA_FROM_FLASH 0x000FF980

#define MEM_DL_STATIC (EVE_RAM_G_SIZE - 4096) /* 0xff000 - start-address of the static part of the display-list, upper 4k of gfx-mem */

uint32_t num_dl_static = 0; /* amount of bytes in the static part of our display-list */
uint8_t tft_active = 0;

typedef struct {
 uint8_t index;
 uint8_t tag;
 uint32_t addressFlash;
 uint32_t addressRamg;
 uint16_t w;
 uint16_t h;
 uint16_t format;
 } flash_Obj;

#define FLASH_ITEMS 5
 /*
  * .edf file:
  * Name                                                       : ID : Offset : Size  : Compression : Type : Sub-Type : Width : Height
  * unified.blob                                               : 0  : 0      : 4096  : 0           : 1    : 0        : 0     : 0
  * Main_304x304_304x304_COMPRESSED_RGBA_ASTC_8x8_KHR.raw      : 1  : 4096   : 23104 : 0           : 2    : 37815    : 304   : 304
  * Main_mini_200x120_200x120_COMPRESSED_RGBA_ASTC_8x8_KHR.raw : 2  : 27200  : 6016  : 0           : 2    : 37815    : 200   : 120
  *	Main_304x304_304x304_COMPRESSED_RGBA_ASTC_8x8_KHR.bin      : 3  : 33216  : 6400  : 1           : 2    : 37815    : 304   : 304
  * Main_mini_200x120_200x120_COMPRESSED_RGBA_ASTC_8x8_KHR.bin : 4  : 39616  : 3584  : 1           : 2    : 37815    : 200   : 120
  */

flash_Obj* pic_all_OBJ[FLASH_ITEMS];					//	id	tag?	Flash	  RAMG	 w		 h		format
	flash_Obj BLOB 					 	 				=	{0,		0,		0, 		0,	 0,	 	  0,	EVE_ASTC_8X8};	//	:	0		:	4096
	flash_Obj pic_astc_304x304_raw					 	=	{1,		0,	  4096,		0,	304U,	304U,	EVE_ASTC_8X8};	//	:	4096	:	23104
	flash_Obj pic_mini_200x120_raw						=	{2,		0,	  27200,	0,	200U,	120U,	EVE_ASTC_8X8};	//	:	27200	:	6016
	flash_Obj pic_astc_304x304_bin					 	=	{3,		0,	  33216,	0,	304U,	304U,	EVE_ASTC_8X8};	//	:	33216	:	6400
	flash_Obj pic_mini_200x120_bin						=	{4,		0,	  39616,	0,	200U,	120U,	EVE_ASTC_8X8};	//	:	39616	:	3584

void initStaticBackground(void);

void initStaticBackground(void)
{
    EVE_cmd_dl(CMD_DLSTART); /* Start the display list */
    EVE_cmd_dl(DL_TAG); /* no tag = 0 - do not use the following objects for touch-detection */

    EVE_cmd_bgcolor(0x00c0c0c0UL); /* light grey */

    EVE_cmd_dl(DL_VERTEX_FORMAT); /* set to 0 - reduce precision for VERTEX2F to 1 pixel instead of 1/16 pixel default */


    /* display the picture from RAM_G */
    EVE_color_rgb(WHITE);
    EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
    EVE_cmd_setbitmap(MEM_LAWETA_MINI, EVE_ASTC_8X8, 200U, 120U);
    EVE_cmd_dl(VERTEX2F((EVE_HSIZE-200),(EVE_VSIZE-120)/2));
    EVE_cmd_dl(DL_END);

    /* display the picture from RAM_G, read before from FLASH */
	EVE_color_rgb(WHITE);
	EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
	EVE_cmd_setbitmap(MEM_LAWETA_FROM_FLASH, EVE_ASTC_8X8, 200U, 120U);
	EVE_cmd_dl(VERTEX2F((100), (100)));
	EVE_cmd_dl(DL_END);

			/* display the RAW picture from FLASH */
			EVE_color_rgb(WHITE);
			EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
			EVE_cmd_setbitmap((0x800000 | pic_astc_304x304_raw.addressFlash), pic_astc_304x304_raw.format, pic_astc_304x304_raw.w, pic_astc_304x304_raw.h);
			EVE_cmd_dl(VERTEX2F(0, 0));
			EVE_cmd_dl(DL_END);

			/* display the BIN picture from FLASH */
			EVE_color_rgb(WHITE);
			EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
			EVE_cmd_setbitmap((0x800000 | pic_astc_304x304_bin.addressFlash),pic_astc_304x304_bin.format, pic_astc_304x304_bin.w, pic_astc_304x304_bin.h);
			EVE_cmd_dl(VERTEX2F(0, pic_astc_304x304_raw.h+10));
			EVE_cmd_dl(DL_END);

    EVE_cmd_text(EVE_HSIZE/2, 15, 29, EVE_OPT_CENTERX, "TEST!");

    EVE_cmd_text(10, EVE_VSIZE - 35, 26, 0, "DL-size:");
    EVE_cmd_text(10, EVE_VSIZE - 20, 26, 0, "Freq:");

    EVE_cmd_text(105, EVE_VSIZE - 20, 26, 0, "Hz");

    EVE_execute_cmd();

    num_dl_static = EVE_memRead16(REG_CMD_DL);

    EVE_cmd_memcpy(MEM_DL_STATIC, EVE_RAM_DL, num_dl_static);
    EVE_execute_cmd();
}


void TFT_init(void)
{
    if(E_OK == EVE_init())
    {
        tft_active = 1;

        EVE_memWrite8(REG_PWM_DUTY, 0x40);  /* setup backlight, range is from 0 = off to 0x80 = max */

        EVE_cmd_inflate(MEM_LAWETA_KW, laweta_kw, sizeof(laweta_kw)); /* load picture into gfx-memory and de-compress it */
        HAL_Delay(10);
        EVE_cmd_inflate(MEM_LAWETA_MINI, laweta_mini, sizeof(laweta_mini)); /* load picture into gfx-memory and de-compress it */
        HAL_Delay(10);

        EVE_cmd_flashread(MEM_LAWETA_FROM_FLASH, pic_mini_200x120_raw.addressFlash , 6016); /* Load picture from FLASH to gfx-memory */
        HAL_Delay(10);

        initStaticBackground();
    }
}


uint16_t display_list_size = 0;
/* check for touch events and setup vars for TFT_display() */
void TFT_touch(void)
{
	/* Not using yet */
}

/*
    dynamic portion of display-handling, meant to be called every 20ms or more
*/
void TFT_display(void)
{
    if(tft_active != 0U)
    {
        EVE_start_cmd_burst(); /* start writing to the cmd-fifo as one stream of bytes, only sending the address once */

        EVE_cmd_dl_burst(CMD_DLSTART); /* start the display list */
        EVE_cmd_dl_burst(DL_CLEAR_COLOR_RGB | GREEN); /* set the default clear color to green */
        EVE_cmd_dl_burst(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG); /* clear the screen - this and the previous prevent artifacts between lists, Attributes are the color, stencil and tag buffers */
        EVE_cmd_dl_burst(DL_TAG);

        EVE_cmd_append_burst(MEM_DL_STATIC, num_dl_static); /* insert static part of display-list from copy in gfx-mem */



        /* display a picture from RAM_G */
        EVE_cmd_setbitmap_burst(MEM_LAWETA_KW, EVE_RGB565, 100U, 100U);

        EVE_cmd_dl_burst(DL_SAVE_CONTEXT);
        EVE_cmd_dl_burst(CMD_LOADIDENTITY);

        EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
        EVE_cmd_dl_burst(VERTEX2F(EVE_HSIZE-100, (LAYOUT_Y1)));
        EVE_cmd_dl_burst(DL_END);
        EVE_cmd_dl_burst(DL_RESTORE_CONTEXT);

				/* DISPLAY from FLASH RAW FILE */
				EVE_cmd_setbitmap_burst(0x800000 | pic_mini_200x120_raw.addressFlash, pic_mini_200x120_raw.format, pic_mini_200x120_raw.w, pic_mini_200x120_raw.h);

				EVE_cmd_dl_burst(DL_SAVE_CONTEXT);
				EVE_cmd_dl_burst(CMD_LOADIDENTITY);

				EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
				EVE_cmd_dl_burst(VERTEX2F((EVE_HSIZE-pic_mini_200x120_raw.w)/2, (EVE_VSIZE - pic_mini_200x120_raw.h)/2));
				EVE_cmd_dl_burst(DL_END);
				EVE_cmd_dl_burst(DL_RESTORE_CONTEXT);

				/* DISPLAY from FLASH BIN FILE */
				EVE_cmd_setbitmap_burst(0x800000 | pic_mini_200x120_bin.addressFlash, pic_mini_200x120_bin.format, pic_mini_200x120_bin.w, pic_mini_200x120_bin.h);

				EVE_cmd_dl_burst(DL_SAVE_CONTEXT);
				EVE_cmd_dl_burst(CMD_LOADIDENTITY);

				EVE_cmd_dl_burst(DL_BEGIN | EVE_BITMAPS);
				EVE_cmd_dl_burst(VERTEX2F((EVE_HSIZE-pic_mini_200x120_bin.w)/2, ((EVE_VSIZE - pic_mini_200x120_bin.h)/2)+ pic_mini_200x120_raw.h));
				EVE_cmd_dl_burst(DL_END);
				EVE_cmd_dl_burst(DL_RESTORE_CONTEXT);

        /* print profiling values */
        EVE_color_rgb_burst(WHITE);

        num_profile_a = num_profile_b;
        num_profile_b = HAL_GetTick();
        float freq = 1.0 /(((float)(num_profile_b-num_profile_a))/1000.0);
        EVE_cmd_number_burst(100, EVE_VSIZE - 35, 26, EVE_OPT_RIGHTX, display_list_size); /* number of bytes written to the display-list by the command co-pro */
        EVE_cmd_number_burst(100, EVE_VSIZE - 20, 26, EVE_OPT_RIGHTX|2, freq); /* duration in us of TFT_loop() for the touch-event part */


        EVE_cmd_dl_burst(DL_DISPLAY); /* instruct the co-processor to show the list */
        EVE_cmd_dl_burst(CMD_SWAP); /* make this list active */

        EVE_end_cmd_burst(); /* stop writing to the cmd-fifo, the cmd-FIFO will be executed automatically after this or when DMA is done */
    }
}
```

There are a lot of attempts to display images in the code, because I tried many ways, but none of them work. Could you please tell me what I am doing wrong?

Thank you in advance.

**Comment by RudolphRiedel** [comment]
Ok, as a baseline I just went thru the necessary steps and modified the minimal Arduino example I uploaded yesterday.
I converted my Pixel-Star image to ASTC 8x8 with EAB, put it in a flash image and wrote it to the RVT50HQBNWC00-B I have on my desk right now.
[PixelStar_ASTC.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11140113/PixelStar_ASTC.zip)

This is the updated example code:
[EVE_Test.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11140126/EVE_Test.zip)

The EVE code is this:
`    if(E_OK == EVE_init()) /* make sure the init finished correctly */
    {
        if(E_OK == EVE_init_flash())
        {
            EVE_cmd_dl(CMD_DLSTART); /* instruct the co-processor to start a new display list */
            EVE_cmd_dl(DL_CLEAR_COLOR_RGB | 0xffffff); /* set the default clear color to white */
            EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);

            EVE_cmd_setbitmap(0x800000 | (4096U/32U), EVE_ASTC_8X8,  100U, 100U);
            EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
            EVE_cmd_dl(VERTEX2F(100, 100));
            EVE_cmd_dl(DL_END);

            EVE_color_rgb(0x000000); /* set the color to black */
            EVE_cmd_text(EVE_HSIZE/2, EVE_VSIZE/2, 30, EVE_OPT_CENTER, "Hello, World!");
            EVE_cmd_dl(DL_DISPLAY); /* mark the end of the display-list */
            EVE_cmd_dl(CMD_SWAP); /* make this list active */
//        EVE_execute_cmd(); /* wait for EVE to be no longer busy */
        }
    }
`

And it displays the image from flash.

>Not only that, but I have some questions:
>Should I use compressed .bin files or .raw to upload on FLASH?

You need to use .raw as the .bin files are zlib compressed after the conversion to ASTC.
The compression makes more sense when you convert to a different format and use CMD_INFLATE2 to
decompress from flash to RAM_G.
Or when you store the data in your controllers memory.

>-Why user pink-fr3ud in discussion https://github.com/RudolphRiedel/FT800-FT813/discussions/36 use dividing by 32 the flash address:
>EVE_cmd_setbitmap_burst(0x800000 | (gauge.addressFlash / 32), gauge.format, gauge.w, gauge.h);

This is correct and due to that CMD_SETBITMAP is using a block address for the flash:
"Source address for bitmap, in RAM_G or flash memory, as a BITMAP_SOURCE parameter.
it shall be in terms of block unit (each block is 32 bytes) when it is located in flash memory"

Going over your source you are missing a call to EVE_init_flash() to put the flash in state full:
"Direct rendering of ASTC based bitmaps from flash is only possible in FLASH_STATE_FULL"

Add that to TFT_init() and fix the address issue and it should work.

**Comment by grados73** [comment]
Thank you for such a quick and detailed reply. 
You are right, when I added flash init and change address of flash memory everything works ok.

Unfortunately it is not finish of my problems, I have one more with touch. Code now looks like:
```
flash_Obj* pic_all_OBJ[FLASH_ITEMS];		//	id	  tag?		Flash	  RAMG	 	w		 h	 		x		y		format
	flash_Obj BLOB 					 	 	=	{0,		0,		0, 			0,	 	0,	 	0,			0,		0,		EVE_ASTC_8X8};	//	:	0		:	4096
	flash_Obj Main_800x480					=	{1,		0,		4096U,		0,		800U,	480U,		0,		0,		EVE_ASTC_8X8};	//	:	4096	:	96000
	flash_Obj kogut_off_rys					= 	{2,		1,		100096U,	0,		112U,	32U,		180, 	40, 	EVE_ASTC_8X8};	// 	:	100096 	: 	896
	flash_Obj kogut_on_rys					= 	{3,		2,		100992U,	0,		112U,	32U,		180, 	40, 	EVE_ASTC_8X8};	//  :   100992  :   896
	flash_Obj kogut_off_button				= 	{4,		3,		101888U,	0,		96U,	80U,		220, 	350, 	EVE_ASTC_8X8};	// 	: 	101888 	:	1920
	flash_Obj kogut_on_button				= 	{5,		4,		103808U,	0,		96U,	80U,		220, 	350, 	EVE_ASTC_8X8};	// 	: 	101888 	:	1920

void drawObjectOnDisplayFomFlashStart(void)
{
	EVE_cmd_dl(CMD_DLSTART); /* instruct the co-processor to start a new display list */
	EVE_cmd_dl(DL_VERTEX_FORMAT); /* set to 0 - reduce precision for VERTEX2F to 1 pixel instead of 1/16 pixel default */
	EVE_cmd_dl(DL_CLEAR_COLOR_RGB | 0x0); /* set the default clear color to black */
	EVE_cmd_dl(DL_CLEAR | CLR_COL | CLR_STN | CLR_TAG);
}

void drawObjectOnDisplayFomFlashFinish(void)
{
	EVE_cmd_dl(DL_END);
	EVE_cmd_dl(DL_DISPLAY); /* mark the end of the display-list */
	EVE_cmd_dl(CMD_SWAP); /* make this list active */
	EVE_execute_cmd(); /* wait for EVE to be no longer busy */
}

void drawObjectOnDisplayFomFlash(flash_Obj object, uint16_t x, uint16_t y)
{
	EVE_cmd_dl(DL_TAG+object.tag);
	EVE_cmd_setbitmap(0x800000 | (object.addressFlash / 32U), object.format, object.w,	object.h);
	EVE_cmd_dl(DL_BEGIN | EVE_BITMAPS);
	EVE_cmd_dl(VERTEX2F(x, y));
}

void initStaticBackground(void)
{
    EVE_cmd_dl(CMD_DLSTART); /* Start the display list */
    EVE_cmd_dl(DL_TAG); /* no tag = 0 - do not use the following objects for touch-detection */
    EVE_cmd_bgcolor(BLACK); /* light grey */
    EVE_cmd_dl(DL_END);
    EVE_cmd_dl(DL_DISPLAY); /* mark the end of the display-list */
    EVE_cmd_dl(CMD_SWAP); /* make this list active */

    drawObjectOnDisplayFomFlashStart();
    drawObjectOnDisplayFomFlash(Main_800x480, Main_800x480.x, Main_800x480.y);
	drawObjectOnDisplayFomFlash(kogut_on_button, kogut_on_button.x, kogut_on_button.y);
    drawObjectOnDisplayFomFlashFinish();
}


void TFT_init(void)
{
    if(E_OK == EVE_init())
    {
        tft_active = 1;

        EVE_memWrite8(REG_PWM_DUTY, 0x40);  /* setup backlight, range is from 0 = off to 0x80 = max */

        if(E_OK == EVE_init_flash())
        {
        	tft_flash_active = 1;
        	initStaticBackground();
        }
    }
}

uint8_t tag;
uint32_t touch_xy;
uint8_t touch_result = 0;
/* check for touch events and setup vars for TFT_display() */
void TFT_touch(void)
{
	if (EVE_busy()) /* is EVE still processing the last display list? */
	{
		return;
	}
	tag = EVE_memRead8(REG_TOUCH_TAG); 			/* read the value for the first touch point */
	touch_xy =  EVE_memRead32(REG_TOUCH_RAW_XY);

	if(tag > 0)
	{
		touch_result = 1;
		HAL_Delay(1);
	}
	if(touch_xy != 0xFFFFFFFF)
	{
		touch_result = 1;
		HAL_Delay(1);
	}
}

/*
    dynamic portion of display-handling, meant to be called every 20ms or more
*/

void TFT_display(void)
{
    if(tft_active != 0U)
    {
    	if(touch_result != 0)
    	{
    		drawObjectOnDisplayFomFlashStart();
			drawObjectOnDisplayFomFlash(Main_800x480, Main_800x480.x, Main_800x480.y);
			drawObjectOnDisplayFomFlash(kogut_off_button, kogut_off_button.x, kogut_off_button.y);
			EVE_cmd_dl_burst(DL_TAG + 20);
			EVE_cmd_button((EVE_HSIZE - 100), (10), 90, 50, 29, 0, "OK");
			drawObjectOnDisplayFomFlashFinish();
    	}
    	else
    	{
    		drawObjectOnDisplayFomFlashStart();
			drawObjectOnDisplayFomFlash(Main_800x480, Main_800x480.x, Main_800x480.y);
			drawObjectOnDisplayFomFlash(kogut_on_button, kogut_on_button.x, kogut_on_button.y);
			EVE_cmd_dl_burst(DL_TAG+20);
			EVE_cmd_button((EVE_HSIZE - 100), (10), 90, 50, 29, 0, "Touch");
			drawObjectOnDisplayFomFlashFinish();
    	}

    }
}
```

I tried using touch on both graphics (giving them tags:1, 2, 3, 4) and adding a separate button (tag=20), but none of them work.
Conditions:
`tag = EVE_memRead8(REG_TOUCH_TAG); 			/* read the value for the first touch point */
	touch_xy =  EVE_memRead32(REG_TOUCH_RAW_XY);

	if(tag > 0)
	{
		touch_result = 1;
		HAL_Delay(1);
	}
	if(touch_xy != 0xFFFFFFFF)
	{
		touch_result = 1;
		HAL_Delay(1);
	}
	`
	
Are never met.
I read https://github.com/RudolphRiedel/FT800-FT813/discussions/57
I checked supply. I am using two different sources of supply, 5V for backlight and 3,36V for logic supply.
Furthermore, I reduce speed of SPI (I tried 10MHz, 5MHz and 1MHz) and I am using maximum output speed of GPIO "LOW".
Moreover, I tried your function "touch_calibrate()", on both "EVE_cmd_calibrate()" and "EVE_calibrate_manual(EVE_HSIZE, EVE_VSIZE)", but I can not go farther than "Please tap on the dot."
In https://github.com/RudolphRiedel/FT800-FT813/discussions/57 [ADTL](https://github.com/ADTL) wrote: 

> I wrote to Riverdi support with the serial numbers of the displays, as a result I received a RMA number and I will send the displays back to them. We'll see how it ends.

Have you encountered it more often that someone has hardware issues with displays from riverdi?

Do you recommend trying to change the configure registers?

I ride in DS that my display RVT50HQBNWC00-B : 
-9th sign : _TOUCH PANEL: C – With Capacitive Touch Panel_, 
so should I read **REG_CTOUCH_TAG** and **REG_CTOUCH_TOUCH_XY**  instead of **REG_TOUCH_TAG** and **REG_TOUCH_RAW_XY**?

I tried it changing TFT_touch() to:
```
uint8_t tag;
uint32_t touch_xy;
uint32_t ct_mode;
uint32_t ct_extend;
uint8_t touch_result = 0;
/* check for touch events and setup vars for TFT_display() */
void TFT_touch(void)
{
	if (EVE_busy()) /* is EVE still processing the last display list? */
	{
		return;
	}
//	tag = EVE_memRead8(REG_TOUCH_TAG); 			/* read the value for the first touch point */
//	touch_xy =  EVE_memRead32(REG_TOUCH_RAW_XY);

	tag = EVE_memRead8(REG_CTOUCH_TAG); /* 0x0030212CUL */
	touch_xy =  EVE_memRead32(REG_CTOUCH_TOUCH0_XY); /* 0x00302124UL */
	ct_mode =  EVE_memRead32(REG_CTOUCH_MODE); /* 0x00302104UL */
	ct_extend =  EVE_memRead32(REG_CTOUCH_EXTENDED); /* 0x00302108UL */

	if(tag > 0)
	{
		touch_result = 1;
		HAL_Delay(1);
	}
	if(touch_xy != 0x80008000)
	{
		touch_result = 1;
		HAL_Delay(1);
	}
}
```
And it still does not work:
![image](https://user-images.githubusercontent.com/71728314/229825269-a14eed18-46ee-484b-a771-00dfd21aceac.png)
REG_CTOUCH_MODE = 3 (0b’11: On mode.)
REG_CTOUCH_EXTEND = 1 (1: Compatibility mode, single touch detection mode)

Have you got any idea what I am doing wrong? Thank you in advance.

**Reply by RudolphRiedel** [reply]
Before going deeper into the details, do you have a logic analyzer to check what is actually happening on the SPI?

**Reply by grados73** [reply]
Yes, I have one basic logic analyzer. What I should check?

**Reply by RudolphRiedel** [reply]
I am looking for what EVE actually answers in a SPI transfer.

I just threw together a simple test and it works, at least with an UNO.
I connect an ESP32 now and if this works I'll check your post.
[EVE_pic_from_flash_touch_Arduino_PlatformIO.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11150693/EVE_pic_from_flash_touch_Arduino_PlatformIO.zip)

**Reply by RudolphRiedel** [reply]
Ok, it works with the ESP32.

**Reply by grados73** [reply]
This is answear on:
-EVE_busy()
![image](https://user-images.githubusercontent.com/71728314/229859388-79cc02bb-a2b2-4a4b-b8ce-e93db588aecc.png)

-EVE_memRead8(REG_TOUCH_TAG);
![image](https://user-images.githubusercontent.com/71728314/229859692-8cc29963-c710-4bc6-a1d9-93c59018a8bb.png)


-EVE_memRead32(REG_TOUCH_RAW_XY);
![image](https://user-images.githubusercontent.com/71728314/229860083-c9a45084-0177-47a2-8527-bb5f8996de2f.png)

**Reply by grados73** [reply]
-EVE_memRead8(REG_CTOUCH_TAG); /* 0x0030212CUL */
![image](https://user-images.githubusercontent.com/71728314/229860690-8f5567e8-ca2d-4403-9082-5092562ff66c.png)

-EVE_memRead32(REG_CTOUCH_TOUCH0_XY); /* 0x00302124UL */
![image](https://user-images.githubusercontent.com/71728314/229860987-59966c1b-970c-49c6-be74-3a752ca2092f.png)

**Reply by RudolphRiedel** [reply]
This looks fine.
Well, except that there is nothing going on for the touch.

I modified my example code with your snippets:
[EVE_Test_touch_Arduino_PlatformIO.zip](https://github.com/RudolphRiedel/FT800-FT813/files/11151223/EVE_Test_touch_Arduino_PlatformIO.zip)


And there are a few issues.
Your initStaticBackground() does not really do anything so I commented it out.
The touch_calibrate() I left in there also is not doing anything so I commented it out as well.

I cheated on your drawObjectOnDisplayFomFlash() function as I only have my pixel star in flash.

The touch did work but it locked up, so I added an "else" to reset touch_result back to zero
and commented out the if(touch_xy).
Oh yes, I also commented out the HAL_Delay(1) - does not seem to do serve a usefull purpose?

In TFT_display() is a bug that I did not catch immediately either.
You have
EVE_cmd_dl_burst(DL_TAG + 20);
in there.

Mixing the _burst Funktions in a non-burst setting does not work as these functions only do
the absolut minimum for maximum performance.

In case of ESP32 what gets executed is:
EVE_dma_buffer[EVE_dma_buffer_index++] = data;

So the crash that made me aware of the issue is caused by writing beyond the 4k DMA buffer.

If you use EVE_start_cmd_burst(); / EVE_end_cmd_burst(); around the display list you can
mix _burst and non-burst functions as the regular functions do support burst-mode with a minor performance penalty.

Apart from crashing after a couple of seconds this has no impact the touch though.

The last thing I added was printing the tag and touch_xy values on the screen.

>I checked supply. I am using two different sources of supply, 5V for backlight and 3,36V for logic supply.

The adapter board I am using is supplying 3.4V for both the backlight and the logic.

>Furthermore, I reduce speed of SPI (I tried 10MHz, 5MHz and 1MHz) and I am using maximum output speed of GPIO "LOW".

I had to check in EVE_cpp_target.cpp what is setup and not using DMA I am running at 10MHz.

>Moreover, I tried your function "touch_calibrate()", on both "EVE_cmd_calibrate()" and "EVE_calibrate_manual(EVE_HSIZE, EVE_VSIZE)",
>but I can not go farther than "Please tap on the dot."

That is worrying since with EVE_cmd_calibrate() the whole touch handling is done by EVE for the calibration.

>Have you encountered it more often that someone has hardware issues with displays from riverdi?

No and all Riverdi displays I have are doing just fine, at least in single-touch mode.

>I ride in DS that my display RVT50HQBNWC00-B :
>-9th sign : TOUCH PANEL: C – With Capacitive Touch Panel,
>so should I read REG_CTOUCH_TAG and REG_CTOUCH_TOUCH_XY instead of REG_TOUCH_TAG and REG_TOUCH_RAW_XY?

No, there is no REG_CTOUCH_TAG and your definition uses the same address as REG_TOUCH_TAG.
By default the mode is single-touch.
You could swith to multi-touch using:
EVE_memWrite8(REG_CTOUCH_EXTENDED, 0);
And that needs to be done after calibration.

Then you could read REG_TOUCH_TAG for the first touch, REG_TOUCH_TAG1 for the second and so on.

>Have you got any idea what I am doing wrong?

Unfortunately not, if the above does not have any side-effects for some reason I have no further suggestions.

**Reply by grados73** [reply]
> I also commented out the HAL_Delay(1) - does not seem to do serve a usefull purpose?

I only used it during debugging, otherwise, when there is nothing in the conditions, I can't put a breakpoint there. Yes, it only has a development purpose.

> In case of ESP32 what gets executed is:
EVE_dma_buffer[EVE_dma_buffer_index++] = data;

I don't use DMA for SPI, I tried it in the beginning, but I decided that it might not work so well and is not needed in my simply purpose, so I use SPI without them.

> In TFT_display() is a bug that I did not catch immediately either.
You have
EVE_cmd_dl_burst(DL_TAG + 20);
in there.

Oh, very stupid bug. Firstly, I wrote everything using _burst, but then I thought that it might be source of problems, so change it to without _burst, but unfortunately forget about this command.

> Then you could read REG_TOUCH_TAG for the first touch, REG_TOUCH_TAG1 for the second and so on.

Ok, thanks, now I understood it.

> Unfortunately not, if the above does not have any side-effects for some reason I have no further suggestions.

I tried changes from your uploaded code, but it's nothing change.
I'm a little unsure what to try anymore, I'll write to riverdi, maybe it will turn out that they have some defective series, for example. Although I once wrote an email to their support and never got a response.... Otherwise, the only thing left for me is to buy a new display and try to run this on it, because a hardware problem is most likely. 

Nevertheless, thank you very much for your time and help. If the situation finds its end I will definitely write here, maybe it will be useful to someone someday. I learned a lot by reading the discussions in this repository.

**Reply by RudolphRiedel** [reply]
I like the puzzle and yes, this is hopefully usefull to others. :-)

One thing I can add is the information for my module.
RVT50HQBNWC00-B
LOT:22/1271
S/N:0011/00018
DATE:2222

Oh, note to myself, this one should not have the multitouch bug anymore, check this. :-)

**Reply by RudolphRiedel** [reply]
Oh, one more, I just dug out my Nucleo-F446RE and connected it, at least for Arduino this works as well.
That ESP32 really was a weird crossover now looking back. :-)

**Reply by grados73** [reply]
This just make me sure that it is a hardware problem with the display, I will check it by ordering a second one.

**Comment by grados73** [comment]
Sorry, I accidentally closed the discussion and I can't change it without adding another post.

**Comment by grados73** [comment]
Yes, it was a hardware problem. I ordered a second display and now it works fine. But it probably shouldn't work without fixing the errors with your help. Thank you.

**Comment by RudolphRiedel** [comment]
Thank you for letting me know, at least this is a solution.
I do wonder though what could possibly kill the touch-interface.
